### DISCUSSION TITLE
Go layer - cannot run `main` in the way VSCode does

### DISCUSSION THREAD

**Discussion by stradicat** [syl20bnr/spacemacs#14808] [Q&A] [answered]
I'm doing a simple exercise in _Go_, which is comprised of 2 files:
- A `main.go` file that runs the program
- A `deck.go` file that holds a custom type `deck`, and a function called `print()` to be called from `main`.

So far, nothing unseen in other programming languages.

`main.go`:
```go
func main(){
    variable := deck{"element1", "element2"} // if correctly imported, `deck` is a mnemonic type alias for `[]string`
    
    cards.print()
}
```

`deck.go`:
````go
import "fmt"

type deck []string // now 'deck' is a type alias for `[]string`
    func (x deck) string {
        for i, card := range d { // for index and element ’card’ in all indexes in slice ‘cards’, print index and element
	fmt.Println(i, card)
     }
````
|

However, when hitting `M-RET x x` to invoke `go-run-main`, I get an error message:
```sh
# command line arguments
./main.go:8:11: undefined: deck
```

This may happen because `go-run-main` is running `go run main.go`, without additional files.
Thing is, for a correct output, what should be run is `go run main.go deck.go`, as apparently VSCode does, since the latter shows the expected output: a printout of elements via the mentioned `print()` function above.

My question is: how can the `go-run-main` be customized, so that additional files can be added (or detected) as parameters?

**Comment by lebensterben** [comment]
you should ask this question in upstream, go-mode

**Reply by stradicat** [reply]
Will do, thanks!

**Reply by stradicat** [reply]
According to the `go-mode` maintainer, the `go-run-main` function I point to as the culprit is not defined in `go-mode`, so it likely belongs to the _Go_ layer in _Spacemacs_. This is a supposition. 
https://github.com/dominikh/go-mode.el/issues/387

**Reply by lebensterben** [reply]
You reported issue with `go-run-main` which is namespaces as a function of `go`.

You should have used the correct name `spacemacs/go-run-main` https://github.com/syl20bnr/spacemacs/blob/247d059289792513963f6aa481f488f8909ef435/layers/%2Blang/go/funcs.el#L144

**Reply by stradicat** [reply]
Sorry, my bad.

**Reply by lebensterben** [reply]
I may check later.
The quick workaround is to change go-run-args variable.

**Reply by lebensterben** [reply]
a fix is provided here

https://github.com/Microsoft/vscode-go/issues/3096

**Reply by stradicat** [reply, marked as answer]
I think that did it.

In `spacemacs/layers/+lang/go/funcs.el`, line 147:

```elisp
   (format (concat go-run-command " %s %s .")
```

Changing it to 
```elisp
   (format (concat go-run-command " .")
```

did the trick.

Thanks for the heads-up!
